src/openforms/registrations/contrib/objects_api/submission_registration.py

```diff
@@ -339,8 +349,23 @@ class ObjectsAPIV2Handler(ObjectsAPIRegistrationHandler[RegistrationOptionsV2]):
     def get_object_data(
         self, submission: Submission, options: RegistrationOptionsV2
     ) -> dict[str, Any]:
+
         state = submission.load_submission_value_variables_state()
         dynamic_values = state.get_data()
+
+        # For every file upload component, we alter the value of the variable to be
+        # the Document API URL(s).
+        for key in dynamic_values.keys():
+            objects_api_attachments = ObjectsAPISubmissionAttachment.objects.filter(
+                submission_file_attachment__submission_variable__submission=submission,
+                submission_file_attachment__submission_variable__key=key,
+            )
+            if objects_api_attachments.exists():
+                urls = list(
+                    objects_api_attachments.values_list("documents_url", flat=True)
+                )
+                dynamic_values[key] = urls[0] if len(urls) == 1 else urls
+
```

Comment:
Would it be better to get all the attachments for the submission, annotate them with the variable key, and then make them into a dictionary of key: attachment. Then iterate over the attachment and with the key update the corresponding variables? I think it might save queries for the variables that are not files?

Reply:
let's optimize this indeed, a `defaultdict(list)` should work well for this. You also need to keep track of the `component["multiple"]` and not determine the type based on the amount of files, since you can have a multiple-allowed field with just the single upload, which will be mapped to an array in the json schema. The current code would write a string instead of an array.

This is also a test case that should be added :-)